sound/soc/sof/ipc.c

```diff
@@ -287,19 +287,34 @@ static void ipc_tx_next_msg(struct work_struct *work)
 		container_of(work, struct snd_sof_ipc, tx_kwork);
 	struct snd_sof_dev *sdev = ipc->sdev;
 	struct snd_sof_ipc_msg *msg;
+	int ret;
 
 	spin_lock_irq(&sdev->ipc_lock);
 
 	/* send message if HW read and message in TX list */
-	if (list_empty(&ipc->tx_list) || !snd_sof_dsp_is_ready(sdev))
+	if (list_empty(&ipc->tx_list) || !snd_sof_dsp_is_ipc_ready(sdev))
 		goto out;
 
 	/* send first message in TX list */
 	msg = list_first_entry(&ipc->tx_list, struct snd_sof_ipc_msg, list);
-	list_move(&msg->list, &ipc->reply_list);
-	snd_sof_dsp_send_msg(sdev, msg);
+	ret = snd_sof_dsp_send_msg(sdev, msg);
+	if (ret < 0) {
+
+		/*
+		 * if ipc tx fails, the msg will be retained in the msg_list,
+		 * so that it can be re-tried until it succeeds or times-out.
+		 */
+		dev_err_ratelimited(sdev->dev,
+				    "error: ipc tx failed with error %d\n",
+				    ret);
+	} else {
+
+		/* message sent. move it to the reply list */
+		list_move(&msg->list, &ipc->reply_list);
+
+		ipc_log_header(sdev->dev, "ipc tx", msg->header);
+	}
 
```

Comment:
I have concern about the logic here:
1. all our send_msg() ops currently won't fail, it only write IPC registers and then return 0, this check here looks superfluous here.
2. even the send_msg() may fail in future(e.g. some send_msg() ops changed), only keep the message in the msg_list is not enough to make sure the message will be re-tried and sent in time(at worst it may need wait next IPC sent e.g. at 1 Second later and then this not-sent IPC may wait timeout actually). We need something like this at the fail case:

```
	/* re-schedule and retry sending messages... */
	snd_sof_ipc_msgs_tx(sdev);
```
As I can't imagine how will send_msg() ops fail, and I am unclear about if the retry at once make sense(or how many times should we retry)...

According to those above, I don't like to send_msg() ops fail check at the moment -- we can't make sure the logic here work.

Reply:
@keyonjie not sure you read the other comments earlier. Pierre made it clear that just because it doesnt fail on our platforms it doesnt mean it will never fail on other platforms

Reply:
About 2. i think it is quite clear that there is not guarantee that it will ever be re-tried. I cant retry with a snd_sof_ipc_msgs_tx() because it will end up in a infinite loop. when do you stop re-scheduling?

Reply:
@keyonjie i dont want to implement a half-baked solution in a hurry. How about I state the limitation here and mention that the IPC re-try mechanism needs to be reworked?

Reply:
@ranj063 yep, that's good to me.